src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs

```diff
@@ -66,69 +101,108 @@ public static Task<int> ReceiveAsync(this Socket socket, ArraySegment<byte> buff
             SocketFlags socketFlags, 
             EndPoint remoteEndPoint)
         {
+            Tuple<Socket, EndPoint> packedArguments = Tuple.Create(socket, remoteEndPoint);
+
             return Task<SocketReceiveFromResult>.Factory.FromAsync(
-                (callback, state) => socket.BeginReceiveFrom(
-                    buffer.Array, 
-                    buffer.Offset, 
-                    buffer.Count, 
-                    socketFlags, 
-                    ref remoteEndPoint, 
-                    callback, 
-                    state),
-                asyncResult => {
-                    int bytesReceived = socket.EndReceiveFrom(asyncResult, ref remoteEndPoint);
-
-                    return new SocketReceiveFromResult() {
+                (targetBuffer, flags, callback, state) =>
+                {
+                    var arguments = (Tuple<Socket, EndPoint>)state;
+                    Socket s = arguments.Item1;
+                    EndPoint e = arguments.Item2;
+
+                    return s.BeginReceiveFrom(
+                        targetBuffer.Array,
+                        targetBuffer.Offset,
+                        targetBuffer.Count,
+                        flags,
+                        ref e,
```

Comment:
This doesn't look correct.  This ref is changing the local e, which is going to be a copy of what's in the tuple, so the end delegate is going to potentially using the wrong e, right?

Same in the later similar transformations.

If I'm right, this also suggests there's a test gap.